lib/asyn-thread.c

```diff
@@ -536,7 +536,7 @@ void Curl_resolver_kill(struct Curl_easy *data)
   /* If we're still resolving, we must wait for the threads to fully clean up,
      unfortunately.  Otherwise, we can simply cancel to clean up any resolver
      data. */
-  if(td && td->thread_hnd != curl_thread_t_null)
+  if(td && td->thread_hnd != curl_thread_t_null && !data->set.may_leak)
```

Comment:
This makes the else block run instead, which isn't what you want is it?

Reply:
This is what I want: see the full code block below.

```c
 if(td && td->thread_hnd != curl_thread_t_null && !data->set.may_leak)
     (void)thread_wait_resolv(data, NULL, FALSE);
   else
     Curl_resolver_cancel(data);
```

As it turns out, the `else` block does call `Curl_resolver_cancel`, which (1) doesn't block and (2) is useful as it avoids a race condition if the thread happens to finish at the same time:

```c
 /* within destroy_async_data() in async_thread.c */

    Curl_mutex_acquire(td->tsd.mtx);
    done = td->tsd.done;
    td->tsd.done = 1;
    Curl_mutex_release(td->tsd.mtx);

    if(!done) {
      Curl_thread_destroy(td->thread_hnd);
    }
*/